plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergUtil.java

```diff
@@ -623,4 +640,99 @@ private static void validateOrcBloomFilterColumns(ConnectorTableMetadata tableMe
             throw new TrinoException(INVALID_TABLE_PROPERTY, format("Orc bloom filter columns %s not present in schema", Sets.difference(ImmutableSet.copyOf(orcBloomFilterColumns), allColumns)));
         }
     }
+
+    public static Optional<String> getLatestMetadataLocation(TrinoFileSystem fileSystem, String location)
+    {
+        Set<String> latestMetadataLocations = new HashSet<>();
+        String metadataDirectoryLocation = format("%s/%s", stripTrailingSlash(location), METADATA_FOLDER_NAME);
+        try {
+            int latestMetadataVersion = -1;
+            FileIterator fileIterator = fileSystem.listFiles(metadataDirectoryLocation);
+            while (fileIterator.hasNext()) {
+                FileEntry fileEntry = fileIterator.next();
+                if (fileEntry.path().endsWith(METADATA_FILE_EXTENSION)) {
+                    OptionalInt version = parseVersion(fileEntry.path());
+                    if (version.isPresent()) {
+                        if (version.getAsInt() > latestMetadataVersion) {
+                            latestMetadataVersion = version.getAsInt();
+                            latestMetadataLocations.clear();
+                            latestMetadataLocations.add(fileEntry.path());
+                        }
+                        else if (version.getAsInt() == latestMetadataVersion) {
+                            latestMetadataLocations.add(fileEntry.path());
+                        }
+                    }
+                }
+            }
+            if (latestMetadataLocations.isEmpty()) {
+                throw new TrinoException(ICEBERG_INVALID_METADATA, "No metadata file exists at location: " + metadataDirectoryLocation);
```

Comment:
we ignore metadata files for which `parseVersion` returns empty

```suggestion
                throw new TrinoException(ICEBERG_INVALID_METADATA, "No versioned metadata file exists at location: " + metadataDirectoryLocation);
```